### ISSUE TITLE
Poor performance with sets if there is a large number data objects

### ISSUE THREAD

**Issue by RKrahl** [icatproject/icat.oaipmh#16]
I did some performance testing for the case that sets are configured and a large number of data objects is disseminated over OAI-PMH.

I setup a test ICAT with 100,000 Investigations. I used the following configuration in `run.properties` to setup two sets:
```
sets = setA setB

sets.setA.name = Example Set A
sets.setA.configurations = inv
sets.setA.condition.inv = pt.name = 'oaipmh-set_a'
sets.setA.join.inv = JOIN a.parameters AS p JOIN p.type AS pt

sets.setB.name = Example Set B
sets.setB.configurations = inv
sets.setB.condition.inv = pt.name = 'oaipmh-set_b'
sets.setB.join.inv = JOIN a.parameters AS p JOIN p.type AS pt
```
I created the corresponding ParameterTypes and InvestigationParameters to mark two random samples of about 50,000 Investigations each as members of each the sets.

I considered different scenarios:

1. I deployed a development version with PR #14 and #15 merged, using the configuration described above

2. I disabled `sets.setA.condition.inv` and `sets.setB.condition.inv` so that all investigations are in both sets.

3. I followed the suggestion in the last comment in #15 and used a hard coded value of 10,000 in the LIMIT clause with the same configuration as in 1.

For scenario 1, it took more then 40 minutes to collect the object ids for the two sets from the ICAT server. Obviously the HTTP connection did time out long before that was done, so the client only got an internal server error.  Scenario 2 and 3 delivered more or less prompt replies, I didn't measure exact timings, though.

**Comment by RKrahl** [comment]
The outcome from scenario 3 suggests that we should indeed query the `maxEntities` value configured in icat.server and use that in the LIMIT clause. That should provide an acceptable performance so that we could release 1.1.0 with that fix. 

Later on, we should do more testing and observe the behavior in production.

**Comment by RKrahl** [comment]
I did some time measurements with the fix provided by #17 in order to check it is working, quantify the effect, and assess the performance in general. I used the same scenario as above, e.g. 100,000 investigations, two sets, approximately 50,000 investigations in each set. I varied the `maxEntities` setting in icat.server and measured the time for icat.oaipmh to reply to a `ListRecords` call. I made the `ListRecords` call with Dublin Core and with DataCite metadata, but there was no significant difference in the timings between both. The client was accessing icat.oaipmh with HTTP (not HTTPS) from the same local network.

Here are the results:

maxEntities | ListRecords
------------|------------
10000 | 4 s
1000 | 24 s
600 | 40 s
100 | 3.6 min
Reference | 550 ms

In the case with `maxEntities=100` I also needed to reduce the `maxResults` setting in icat.oaipmh to 5, otherwise the query for compiling the result list would hit that limit. In all other cases, `maxResults` was left with the default setting of 50. In the last line, I added a reference measurement with `sets.setA.condition.inv` and `sets.setB.condition.inv` disabled so that all investigations are in both sets.

As a summary, the fix clearly has the expected effect. The response time in the case with `maxEntities=10000` is acceptable. So I'd say, with may release icat.oaipmh with that fix. On the other hand, the comparision with the reference clearly shows that the compilation of the lists of set members still adds a significant overhead.
